pkg/ingress/controller.go

```diff
@@ -317,30 +325,47 @@ election:
 }
 
 func (c *Controller) run(ctx context.Context) {
-	log.Infow("controller now is running as leader",
+	log.Infow("controller tries to leading ...",
 		zap.String("namespace", c.namespace),
 		zap.String("pod", c.name),
 	)
+
+	var cancelFunc context.CancelFunc
```

Comment:
It seems that you don't merge from the latest master, here we already have some methods to finalize the context passed to `run`.

Here we cannot cancel the context of `ctx` in parameters, so we cannot give up the leader role by ourselves.

Reply:
```
...
election:
	curCtx, cancel := context.WithCancel(rootCtx)
	c.leaderContextCancelFunc = cancel
	elector.Run(curCtx)
	select {
	case <-rootCtx.Done():
		return nil
	default:
		goto election
	}
...
```
```
func (c *Controller) run(ctx context.Context) {
...
	defer c.leaderContextCancelFunc()
...
	<-ctx.Done()
	c.wg.Wait()
}
```

Not, you mean use one c.leaderContextCancelFunc to control the run context ? It is not enough.
I think the code should wait for all goroutine to exit in c.run before executing the next election.

Reply:
The reason why I introduced the `leaderContextCancelFunc` is, package `leaderelection` doesn't quit the lease renew loop even if `c.run` exits.

> Not, you mean use one c.leaderContextCancelFunc to control the run context ? It is not enough.
I think the code should wait for all goroutine to exit in c.run before executing the next election.

Yep, waiting is required before it exits. This is logical inside `c.run`.

Reply:
The cancelFunc manage context of c.run, leaderContextCancelFun just manage the election context.
If call leaderContextCancelFun in health check the election will run into next loop and not wait last c.run exited.

Reply:
I didn't say we should call `leaderContextCancelFun` in the health check. Just call it before the `c.run` return. Or the leader role cannot be given up.